raft/tracker/inflights.go

```diff
@@ -22,37 +28,40 @@ package tracker
 type Inflights struct {
 	// the starting index in the buffer
 	start int
-	// number of inflights in the buffer
-	count int
 
-	// the size of the buffer
-	size int
+	count int    // number of inflight messages in the buffer
+	bytes uint64 // number of inflight bytes
+
+	size     int    // the max number of inflight messages
+	maxBytes uint64 // the max total byte size of inflight messages
 
-	// buffer contains the index of the last entry
-	// inside one message.
-	buffer []uint64
+	// buffer is a ring buffer containing info about all in-flight messages.
+	buffer []inflight
 }
 
-// NewInflights sets up an Inflights that allows up to 'size' inflight messages.
-func NewInflights(size int) *Inflights {
+// NewInflights sets up an Inflights that allows up to 'size' inflight messages,
+// with the total size up to maxBytes. If maxBytes is 0 then there is no byte
+// size limit. The maxBytes limit is soft, a single message may overflow it.
```

Comment:
Ah, you mention it here. Maybe that is ok. But you could clarify what you mean by "a single message will overflow it". The design of the interface here is that we will always end up overflowing by one message. There's no method on `Inflights` to check whether adding a message would lead to overflow. You add the message and check after.